### PR TITLE
Added HAProxy example

### PR DESCRIPTION
**Pull request by benpro**
Proposal of an example with HAProxy. Asked on #4541 

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

File: docs/reverse_proxy.rst [hunk 1]
```diff
@@ -85,6 +85,28 @@ Let's assume that we expect clients to connect to our server at
           </Location>
       </VirtualHost>
 
+* HAProxy::
+
+      frontend https
+        bind 0.0.0.0:443 v4v6 ssl crt /etc/ssl/haproxy/ strict-sni alpn h2,http/1.1
+        bind :::443 ssl crt /etc/ssl/haproxy/ strict-sni alpn h2,http/1.1
+        
+        # A standard website
```
Comment (richvdh) [inline docs/reverse_proxy.rst]: let's get rid of this? we don't have it for any of the other proxies and I think it's confusing.

File: docs/reverse_proxy.rst [hunk 2]
```diff
@@ -85,6 +85,28 @@ Let's assume that we expect clients to connect to our server at
...
+        acl www hdr(host) -i www.example.com
+        use_backend www if www
+        
+        # Matrix-Synapse
+        acl matrix hdr(host) -i matrix.example.com
+        use_backend matrix if matrix
+        
+      frontend matrix-synapse
```
Comment (richvdh) [inline docs/reverse_proxy.rst]: `matrix-federation` might be a better name

File: docs/reverse_proxy.rst [hunk 3]
```diff
@@ -85,6 +85,28 @@ Let's assume that we expect clients to connect to our server at
...
+
+      frontend https
+        bind 0.0.0.0:443 v4v6 ssl crt /etc/ssl/haproxy/ strict-sni alpn h2,http/1.1
+        bind :::443 ssl crt /etc/ssl/haproxy/ strict-sni alpn h2,http/1.1
+        
+        # A standard website
+        acl www hdr(host) -i www.example.com
+        use_backend www if www
+        
+        # Matrix-Synapse
```
Comment (richvdh) [inline docs/reverse_proxy.rst]: "Matrix-Synapse" doesn't really mean much.

"Matrix client traffic", maybe?

File: docs/reverse_proxy.rst (hunk 1 above)
Comment (benpro) [reply docs/reverse_proxy.rst]: All right.

File: docs/reverse_proxy.rst (hunk 2 above)
Comment (benpro) [reply docs/reverse_proxy.rst]: Ok!

**Pull request merged**